File: go/iam/api_user/v1/api_credentials.go
package api_userv1

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const MESH_API_CREDENTIALS_ENV_VAR = "MESH_API_CREDENTIALS"

// LoadAPICredentialsFromFile loads API APICredentials from a JSON file
func LoadAPICredentialsFromFile(path string) (*APICredentials, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read APICredentials file: %w", err)
	}

	var creds APICredentials
	if err := json.Unmarshal(data, &creds); err != nil {
		return nil, fmt.Errorf("failed to parse APICredentials file: %w", err)
	}

	if creds.ApiKey == "" {
		return nil, fmt.Errorf("api_key is required in APICredentials file")
	}

	if creds.Group == "" {
		return nil, fmt.Errorf("group is required in APICredentials file")
	}

	if !strings.HasPrefix(creds.Group, "groups/") {
		return nil, fmt.Errorf("group must be in format groups/{group_id}, got: %s", creds.Group)
	}

	return &creds, nil
}

// APICredentialsFromEnvironment loads API APICredentials from the file path specified in MESH_API_CREDENTIALS environment variable
func APICredentialsFromEnvironment() (*APICredentials, error) {
	path := os.Getenv(MESH_API_CREDENTIALS_ENV_VAR)
	if path == "" {
		return nil, fmt.Errorf("MESH_API_CREDENTIALS environment variable not set")
	}

	return LoadAPICredentialsFromFile(path)
}

// DefaultCredentialsPath returns the OS-specific default path for Mesh API credentials
func DefaultCredentialsPath() string {
	configDir, err := os.UserConfigDir()
	if err != nil {
		// Fallback to home directory if UserConfigDir fails
		homeDir, err := os.UserHomeDir()
		if err != nil {
			return ""
		}
		configDir = filepath.Join(homeDir, ".config")
	}

	return filepath.Join(configDir, "mesh", "credentials.json")
}

// LoadDefaultCredentials loads API credentials from the default location if the file exists
func LoadDefaultCredentials() (*APICredentials, error) {
	path := DefaultCredentialsPath()
	if path == "" {
		return nil, fmt.Errorf("unable to determine default credentials path")
	}

	// Check if file exists before attempting to load
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return nil, fmt.Errorf("default credentials file does not exist: %s", path)
	}

	return LoadAPICredentialsFromFile(path)
}

// FindCredentials searches for API credentials using the standard discovery hierarchy:
// 1. MESH_API_CREDENTIALS environment variable (if set)
// 2. Default credential file location
func FindCredentials() (*APICredentials, error) {
	// Try environment variable first (existing behavior)
	if creds, err := APICredentialsFromEnvironment(); err == nil {
		return creds, nil
	}

	// Try default file location
	if creds, err := LoadDefaultCredentials(); err == nil {
		return creds, nil
	}

	return nil, fmt.Errorf("no API credentials found: tried environment variable (%s) and default location (%s)",
		MESH_API_CREDENTIALS_ENV_VAR, DefaultCredentialsPath())
}

File: python/src/meshtrade/iam/api_user/v1/api_credentials.py
"""
API Credentials loading functionality for Meshtrade SDK.
"""

import json
import os

from .api_credentials_pb2 import APICredentials

MESH_API_CREDENTIALS_ENV_VAR = "MESH_API_CREDENTIALS"


def load_api_credentials_from_file(path: str) -> APICredentials:
    """Load API credentials from a JSON file.

    Args:
        path: Path to the credentials JSON file

    Returns:
        APICredentials object with loaded credentials

    Raises:
        FileNotFoundError: If the credentials file doesn't exist
        ValueError: If the credentials file is invalid or missing required fields
    """
    try:
        with open(path) as f:
            data = json.load(f)
    except FileNotFoundError as e:
        raise FileNotFoundError(f"Failed to read API credentials file: {path}") from e
    except json.JSONDecodeError as e:
        raise ValueError(f"Failed to parse API credentials file: {e}") from e

    api_key = data.get("api_key")
    group = data.get("group")

    if not api_key:
        raise ValueError("api_key is required in API credentials file")

    if not group:
        raise ValueError("group is required in API credentials file")

    if not group.startswith("groups/"):
        raise ValueError(f"group must be in format groups/{{group_id}}, got: {group}")

    creds = APICredentials()
    creds.api_key = api_key
    creds.group = group

    return creds


def api_credentials_from_environment() -> APICredentials | None:
    """Load API credentials from the file path specified in MESH_API_CREDENTIALS environment variable.

    Returns:
        APICredentials object if environment variable is set and file exists, None otherwise

    Raises:
        ValueError: If the credentials file is invalid
    """
    path = os.getenv(MESH_API_CREDENTIALS_ENV_VAR)
    if not path:
        return None

    return load_api_credentials_from_file(path)


def default_credentials_path() -> str:
    """Return the OS-specific default path for Mesh API credentials.
    
    Returns:
        Default credential file path based on OS conventions:
        - Linux/macOS: $HOME/.config/mesh/credentials.json  
        - Windows: %APPDATA%/mesh/credentials.json
    """
    if os.name == 'nt':  # Windows
        app_data = os.environ.get('APPDATA')
        if app_data:
            return os.path.join(app_data, 'mesh', 'credentials.json')
        # Fallback if APPDATA not set
        return os.path.expanduser('~/.config/mesh/credentials.json')
    else:  # Linux/macOS/Unix
        # Try XDG_CONFIG_HOME first, fallback to ~/.config
        config_home = os.environ.get('XDG_CONFIG_HOME')
        if config_home:
            return os.path.join(config_home, 'mesh', 'credentials.json')
        return os.path.expanduser('~/.config/mesh/credentials.json')


def load_default_credentials() -> APICredentials | None:
    """Load API credentials from the default location if the file exists.

    Returns:
        APICredentials object if default file exists and is valid, None otherwise

    Raises:
        ValueError: If the credentials file exists but is invalid
    """
    path = default_credentials_path()

    # Check if file exists before attempting to load
    if not os.path.isfile(path):
        return None

    return load_api_credentials_from_file(path)


def find_credentials() -> APICredentials | None:
    """Search for API credentials using the standard discovery hierarchy.
    
    Discovery order:
    1. MESH_API_CREDENTIALS environment variable (if set)
    2. Default credential file location
    
    Returns:
        APICredentials object if found using any method, None if no credentials found

    Raises:
        ValueError: If a credentials file is found but is invalid
    """
    # Try environment variable first (existing behavior)
    creds = api_credentials_from_environment()
    if creds:
        return creds

    # Try default file location
    creds = load_default_credentials()
    if creds:
        return creds

    return None
